src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs

```diff
@@ -0,0 +1,146 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Abstractions.Installer;
+using Microsoft.TemplateEngine.Edge.Settings;
+using Newtonsoft.Json;
+
+namespace Microsoft.TemplateEngine.Edge.BuiltInManagedProvider
+{
+    internal sealed class GlobalSettings : IGlobalSettings, IDisposable
+    {
+        private readonly Paths _paths;
+        private readonly IEngineEnvironmentSettings _environmentSettings;
+        private readonly string _globalSettingsFile;
+        private IDisposable? _watcher;
+        private bool _locked;
+        private volatile bool _disposed;
+
+        public GlobalSettings(IEngineEnvironmentSettings environmentSettings, string globalSettingsFile)
+        {
+            _environmentSettings = environmentSettings ?? throw new ArgumentNullException(nameof(environmentSettings));
+            _globalSettingsFile = globalSettingsFile ?? throw new ArgumentNullException(nameof(globalSettingsFile));
+            _paths = new Paths(environmentSettings);
+            environmentSettings.Host.FileSystem.CreateDirectory(Path.GetDirectoryName(_globalSettingsFile));
+            _watcher = environmentSettings.Host.FileSystem.WatchFileChanges(_globalSettingsFile, FileChanged);
+        }
+
+        private void FileChanged(object sender, FileSystemEventArgs e)
+        {
+            SettingsChanged?.Invoke();
+        }
+
+        public event Action? SettingsChanged;
+
+        public async Task<IDisposable> LockAsync(CancellationToken token)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(GlobalSettings));
+            }
+            token.ThrowIfCancellationRequested();
+            // We must use Mutex because we want to lock across different processes that might want to modify this settings file
+            var mutex = await AsyncMutex.WaitAsync($"812CA7F3-7CD8-44B4-B3F0-0159355C0BD5{_globalSettingsFile}".Replace("\\", "_").Replace("/", "_"), token, Unlocked).ConfigureAwait(false);
```

Comment:
`Unlocked` callback will be called by the thread that is owning the mutex and that may be unexpected to the caller. I don't see an implication of this here and someone can unknowingly put additional logic into `Unlocked` method.

Reply:
If you look at how callback is used, its used by GlobalSettings, that doesn't have control over when AsyncMutex is disposed, but needs to be notified about releasing of lock, do you have suggestion on how to solve this?

Reply:
At the very least, shouldn't `GlobalSettings._locked` be volatile? We are setting/reading it from different threads and dirty reads are possible (I guess.).

Reply:
Removed callback and made _locked volitile